accounting/invoice/tests/test_admin.py

```diff
@@ -20,3 +20,116 @@
+                                                 pdf_path='',
+                                                 paid=False, approved=choices.InvoiceApproval.automatically)
+        # Make line items just for invoice 2 and 3, so invoice 1 has a total of '0.00 EUR'.
+        factories.LineItemFactory(invoice=self.invoice2, quantity=1, price__amount=100, price__currency='EUR')
+        factories.LineItemFactory(invoice=self.invoice3, quantity=2, price__amount=100, price__currency='EUR')
+        super().setUp()
+
+    def _refresh_invoices(self):
+        """Refresh the test case's invoices from the database to have the object reflect new updates."""
+        for invoice in [self.invoice1, self.invoice2, self.invoice3]:
+            invoice.refresh_from_db()
+
+    @ddt.data(
+        ('invoice1', '0.00 EUR'),
+        ('invoice2', '100.00 EUR'),
+        ('invoice3', '200.00 EUR'),
+    )
+    @ddt.unpack
+    def test_total(self, invoice, expected_total):
+        """The total is displayed with appropriate cost and currency."""
+        self.assertEqual(self.admin.total(getattr(self, invoice)), expected_total)
+
+    @ddt.data(
+        ('invoice1', '<a href="https://drive.google.com/f/asdf1.pdf">Click here to see PDF.</a>'),
+        ('invoice2', '<a href="https://drive.google.com/f/asdf2.pdf">Click here to see PDF.</a>'),
+        ('invoice3', 'No PDF available.'),
+    )
+    @ddt.unpack
+    def test_pdf(self, invoice, expected_pdf_link):
+        """The PDF link returns the appropriate HTML and URL."""
+        self.assertEqual(self.admin.pdf(getattr(self, invoice)), expected_pdf_link)
+
+    def test_jira_timesheet(self):
+        """The JIRA Timesheet link contains the appropriate URL parameters."""
+        self.provider1.user.username = 'umanshahzad'
+        self.assertEqual(
+            self.admin.jira_timesheet(self.invoice1),
+            '<a href="'
+            'https://jira.atlassian.com/secure/TempoUserBoard!timesheet.jspa'
+            '?userId=umanshahzad'
+            '&amp;periodType=BILLING'
+            '&amp;periodView=DATES'
+            '&amp;from=2017-12-01'
+            '&amp;to=2017-12-31">'
+            'Click here to see JIRA.'
+            '</a>'
+        )
+
+    def test_mark_paid(self):
+        """Marking a set of invoices as paid through the admin should only mark selected invoices as paid."""
+        self.client.force_login(self.admin_user)
+        self.client.post(self.url, data={
+            'action': 'mark_paid',
+            ACTION_CHECKBOX_NAME: [self.invoice1.pk, self.invoice2.pk],
+        })
+
+        self._refresh_invoices()
+        self.assertTrue(self.invoice1.paid)
+        self.assertTrue(self.invoice2.paid)
+        self.assertFalse(self.invoice3.paid)
+
+    def test_mark_approved(self):
+        """Marking a set of invoices as approved through the admin should only mark selected invoices as approved.
+        Approval status remains the same if already previously approved."""
```

Comment:
@itsjeyd The only way to do that would be to either add another invoice or change one of the invoice's original status. Otherwise, since invoice1/2/3 are not/manually/automatically approved respectively, nothing would change from the operation if we left invoice1 out.

I'm going to add another invoice in that case, which will be not approved from the start, and not included in the selection.

Reply:
There was also a really funny bug here which I randomly noticed. You will see it in the next commit ;p